diesel/src/sqlite/connection/statement_iterator.rs

```diff
@@ -78,8 +85,13 @@ where
             Ok(row) => row,
             Err(e) => return Some(Err(e)),
         };
+        if self.column_indices.is_none() {
+            if let Err(e) = self.populate_column_indices() {
+                return Some(Err(e));
+            }
+        }
         row.map(|row| {
-            let row = row.into_named(&self.column_indices);
+            let row = row.into_named(self.column_indices.as_ref().expect("it's there because we populated it above"));
```

Comment:
So this seems to not break memory safety specifically because this call to `into_named` enforces downgrading the lifetime of `'a` in `HashMap<&'a str, usize>` to the lifetime of `self`.
Maybe we could explicit it there in order to contain why this works closer? (It looks that currently it wouldn't be hard to change the signature of `into_named`, see it compile and call it a day.)
```rust
impl<'a, T> Iterator for NamedStatementIterator<'a, T>
where
    T: QueryableByName<Sqlite>,
{
    type Item = QueryResult<T>;

    fn next<'s>(&'s mut self) -> Option<Self::Item> {
        let row = match self.stmt.step() {
            Ok(row) => row,
            Err(e) => return Some(Err(e)),
        };
        if self.column_indices.is_none() {
            if let Err(e) = self.populate_column_indices() {
                return Some(Err(e));
            }
        }
        // explicitly downgrade the lifetime to a correct one
        let column_indices: &'s HashMap<&'s str, usize> = self.column_indices.as_ref().expect("it's there because we populated it above");
        row.map(|row| {
            let row = row.into_named(column_indices);
            T::build(&row).map_err(DeserializationError)
        })
    }
}
```

Reply:
This is not possible, it results in the following error message:
```
error[E0597]: `self` does not live long enough
  --> diesel/src/sqlite/connection/statement_iterator.rs:94:34
   |
83 |     fn next<'s>(&'s mut self) -> Option<Self::Item> {
   |             -- lifetime `'s` defined here
...
93 |         row.map(|row| {
   |                 ----- value captured here
94 |             let indices: &'s _ = self.column_indices.as_ref().expect("it's there because we populated it above");
   |                          -----   ^^^^ borrowed value does not live long enough
   |                          |
   |                          type annotation requires that `self` is borrowed for `'s`
...
98 |     }
   |     - `self` dropped here while still borrowed
```

That written: I'm quite sure no one will actually touch this code again, as this is already rewritten on the master branch in such a way that this cannot happen there. (Though we should check if we could trigger there the multiple calls to `sqlite3_column_name` thing there.)